src/main/kotlin/org/rust/ide/annotator/cargoCheck/CargoCheckAnnotatorPass.kt

```diff
@@ -0,0 +1,165 @@
+
+        val update = object : Update(file) {
+            override fun setRejected() {
+                super.setRejected()
+                doFinish(getHighlights(), modificationStampBefore)
+            }
+
+            override fun run() {
+                if (!documentChanged(modificationStampBefore) && !myProject.isDisposed) {
+                    BackgroundTaskUtil.runUnderDisposeAwareIndicator(myProject, Runnable {
+                        doAnnotate()
+                        ReadAction.run<RuntimeException> {
+                            ProgressManager.checkCanceled()
+                            if (!documentChanged(modificationStampBefore)) {
+                                doApply()
+                                doFinish(getHighlights(), modificationStampBefore)
+                            }
+                        }
+                    })
+                }
+            }
+        }
+
+        factory.scheduleExternalActivity(update)
+    }
+
+    private fun doAnnotate() {
+        val dumbService = DumbService.getInstance(myProject)
+        if (dumbService.isDumb && !DumbService.isDumbAware(CargoCheckAnnotator)) return
+        try {
+            annotationResult = CargoCheckAnnotator.doAnnotate(annotationInfo ?: return)
+        } catch (t: Throwable) {
+            process(t)
+        }
+    }
+
+    private fun doApply() {
+        if (annotationResult == null || !file.isValid) return
+        try {
+            CargoCheckAnnotator.apply(file, annotationResult, annotationHolder)
+        } catch (t: Throwable) {
+            process(t)
+        }
+    }
+
+    private fun doFinish(highlights: List<HighlightInfo>, modificationStampBefore: Long) {
+        val document = document ?: return
+        ApplicationManager.getApplication().invokeLater({
+            if (!documentChanged(modificationStampBefore) && !myProject.isDisposed) {
+                UpdateHighlightersUtil.setHighlightersToEditor(
+                    myProject,
+                    document,
+                    0,
+                    file.textLength,
+                    highlights,
+                    colorsScheme,
+                    id
+                )
+                DaemonCodeAnalyzerEx.getInstanceEx(myProject).fileStatusMap.markFileUpToDate(document, id)
+            }
+        }, ModalityState.stateForComponent(editor.component))
+    }
+
+    private fun getHighlights(): List<HighlightInfo> =
+        annotationHolder.map(HighlightInfo::fromAnnotation)
+
+    private fun documentChanged(modificationStampBefore: Long): Boolean =
+        myDocument?.modificationStamp != modificationStampBefore
+
+    companion object {
+        private val LOG: Logger = Logger.getInstance(CargoCheckAnnotatorPass::class.java)
+
+        private fun process(t: Throwable) {
+            if (t is ProcessCanceledException) throw t
+            LOG.error(t)
+        }
+    }
+}
+
+class CargoCheckAnnotatorPassFactory(
```

Comment:
Let's rename it to `RsCargoCheckAnnotatorPass` as well as all our extension point impls